core/primitives/src/errors.rs

```diff
@@ -0,0 +1,236 @@
+pub enum InvalidTxError {
+    InvalidSigner(AccountId),
+    SignerDoesNotExist(AccountId),
+    InvalidAccessKey(InvalidAccessKeyError),
+    InvalidNonce(Nonce, Nonce),
+    InvalidReceiver(AccountId),
+    InvalidSignature,
+    NotEnoughBalance(AccountId, Balance, Balance),
+    RentUnpaid(AccountId),
+    CostOverflow,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+pub enum InvalidAccessKeyError {
+    AccessKeyNotFound(AccountId, PublicKey),
+    ReceiverMismatch(AccountId, AccountId),
+    MethodNameMismatch(String),
+    ActionError,
+    NotEnoughAllowance(AccountId, PublicKey, Balance, Balance),
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+pub enum ActionError {
+    AccountAlreadyExists(AccountId),
+    AccountDoesNotExist(Action, AccountId),
+    CreateAccountNotAllowed(AccountId, AccountId),
+    ActorNoPermission(AccountId, AccountId, Action),
+    DeleteKeyDoesNotExist(AccountId),
+    AddKeyAlreadyExists(PublicKey),
+    DeleteAccountStaking(AccountId),
+    DeleteAccountHasRent(AccountId, Balance),
+    RentUnpaid(AccountId),
+    TriesToUnstake(AccountId),
+    TriesToStake(AccountId, Balance, Balance, Balance),
+    FunctionCallError(String), // TODO type
+}
+
+impl Display for InvalidTxError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        match self {
+            InvalidTxError::InvalidSigner(signer_id) => {
+                write!(f, "Invalid signer account ID {:?} according to requirements", signer_id)
+            }
+            InvalidTxError::SignerDoesNotExist(signer_id) => {
+                write!(f, "Signer {:?} does not exist", signer_id)
+            }
+            InvalidTxError::InvalidAccessKey(access_key_error) => access_key_error.fmt(f),
+            InvalidTxError::InvalidNonce(tx_nonce, ak_nonce) => write!(
+                f,
+                "Transaction nonce {} must be larger than nonce of the used access key {}",
+                tx_nonce, ak_nonce
+            ),
+            InvalidTxError::InvalidReceiver(receiver_id) => {
+                write!(f, "Invalid receiver account ID {:?} according to requirements", receiver_id)
+            }
+            InvalidTxError::InvalidSignature => {
+                write!(f, "Transaction is not signed with a public key of the signer")
+            }
+            InvalidTxError::NotEnoughBalance(signer_id, balance, cost) => write!(
+                f,
+                "Sender {} does not have enough balance {} for operation costing {}",
+                signer_id, balance, cost
+            ),
+            InvalidTxError::RentUnpaid(signer_id) => {
+                write!(f, "Failed to execute, because the account {} wouldn't have enough to pay required rent", signer_id)
+            }
+            InvalidTxError::CostOverflow => {
+                write!(f, "Transaction gas or balance cost is too high") 
```

Comment:
What does it means by `balance cost`?

Reply:
This error is the same as InvalidTxError::NotEnoughBalance except we cannot compute the cost of the operation but know it doesn't fit in u128.

`IntegerOverflow` doesn't make much sense. The validator took some numbers from my transaction, did some operations with them and got an overflow? What does it mean and how is it my fault?

Reply:
What confuses me is the name. `CostOverflow` name together with covering both cases makes it confusing.. Simple `GasOverflow` and `BalanceOverflow` - makes much more sense.